src/main/java/net/minecraftforge/common/capabilities/AttachCapabilitiesEvent.java

```diff
@@ -0,0 +1,203 @@
+ */
+
+package net.minecraftforge.common.capabilities;
+
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraftforge.common.capabilities.IAttachedCapabilityProvider.ICopyableCapabilityProvider;
+import net.minecraftforge.common.capabilities.IAttachedCapabilityProvider.ICompleteCapabilityProvider;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * The AttachCapabilitiesEvent is used to attach {@link IAttachedCapabilityProvider}s to game objects.<br>
+ * The firing of this event is determined by the provider object, and is not consistent across usages.<br>
+ * For most cases, this is fired whenever the {@link CapabilityDispatcher} must be resolved.<br>
+ * This means that typically, this event is not fired until capabilities are attempted to be accessed for a particular object.
+ */
+public abstract class AttachCapabilitiesEvent<T extends ICapabilityProvider> extends Event
+{
+
+    protected final T obj;
+    protected final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> providers = new IdentityHashMap<>(); // Type objects can be compared using ==
+    protected final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> view = Collections.unmodifiableMap(providers);
+    protected final Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byName = new HashMap<>();
+    protected final Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byNameView = Collections.unmodifiableMap(byName);
+
+    protected AttachCapabilitiesEvent(T obj)
+    {
+        this.obj = obj;
+    }
+
+    /**
+     * Returns the object that is having capabilities attached.
+     */
+    public final T getObject()
+    {
+        return this.obj;
+    }
+
+    /**
+     * Returns the number of successful attachments.
+     */
+    public final int getSize()
+    {
+        return this.providers.size();
+    }
+
+    /**
+     * Adds a {@link IAttachedCapabilityProvider} to this object.<br>
+     * Only one provider may exist for any given {@link CapabilityType}, on a first-come-first-serve basis.<br>
+     * Additionally, the ID of all providers must be unique across the all {@link CapabilityType}s.
+     * 
+     * @param provider The provider being attached to this game object.
+     * @return true, if this provider was successfully attached, otherwise false.
+     * @throws UnsupportedOperationException If a provider with the same ID already exists.
+     */
+    public boolean addCapability(IAttachedCapabilityProvider<?, T> provider)
```

Comment:
So, looking at this, I have a use case where I have an item stack that exposes four separate capabilities through the same object: a fluid handler, an item handler, and a heat handler (a capability unique to my mod). Currently, this looks like this:

```java
static class Instance implements ICapabilityProvider, ICapabilitySerializable, IItemHandler, IFluidHandler, IFluidHandlerItem, IHeatHandler
```
This single instance gets attached to the item stack, and returns all the above capabilities when queried.

Under this new system, (correct me if I'm wrong), this would **not be the case**. I would be required to create a separate `IAttachedCapabilityProvider` for each capability, which then just acts as a proxy to look into the internal capability? Meaning I now am effectively creating my custom capability object `Instance` above, and four other other custom proxy objects which then get attached.

This seems... inconvenient, to say the least. Is there not a way this can support attaching single objects which can provide multiple capabilities without this indirection?

Reply:
No, unfortunately.  Allowing a single object to provide multiple types is basically regressing back to the current system where attached providers have getCap(type, side), which means you have to iterate all of them whenever any cap is asked for.

Reply:
The attached capabilities are still stored by type on the attached object though, are they not? Why can a provider not say "I provide capabilities A, B, C", which is then baked into the by-type map, except now the same instance is used for several capabilities. It doesn't need to be a runtime queryable check of which capabilities are provided.

Reply:
(Note: you probably shouldn't expose both `IFluidHandler` and `IFluidHandlerItem` - only the latter should be used?)

Reply:
I see the following problems with multityping attached providers
* The generic bound <C> has to be removed.
* getType() must return an array instead of a single object.
* The previous issue means that addCapability in AttachCapabilitiesEvent needs a higher level of complexity to attempt to add to multiple maps, and would need more expensive error reporting (rather than true/false success).
* All exposed capabilities are serialized under the same ID, and are not disjoint anymore.

There is a burden of complexity to allow a single attached object to provide multiple caps, and there is a desire to reduce complexity here.

Considering your design there - why do you have one object which is simultaneously an item handler, fluid handler, and heat handler?  Assuming these handlers have no interoperability, I see no reason they should be merged into a single class like this.

Reply:
> Assuming these handlers have no interoperability

*Narrator: These handlers had interoperability.*

I'd still say this is inconvenient as compared to the current system, objectives of this PR notwithstanding. I'd either have to attach proxy capabilities, or be constantly querying capability A from capability B to know the actual state which ends up to be the same thing. If the point is really to "reduce complexity" and just shove that onto the modder, then sure, why not.

Reply:
You can still construct internal shared objects and wrap that object into the attached object, though I suppose the problem therein lies in cloning.  I'll see how complex it actually makes it to share objects like that.  I don't think the generics will like me very much.

Reply:
fcee166